leader/leader.go

```diff
@@ -1,59 +1,47 @@
 package leader
 
 import (
-	"encoding/json"
-	"io/ioutil"
-	"net/http"
-	"os"
-)
-
-const defaultURL = "http://localhost:4040/"
+	"context"
+	"sync/atomic"
 
-type LeaderElector struct {
-	URL string
-}
+	"github.com/hashicorp/go-hclog"
+	operator_leader "github.com/operator-framework/operator-lib/leader"
+	"k8s.io/client-go/kubernetes"
+)
 
-type LeaderResponse struct {
-	Name string `json:"name"`
+type Elector interface {
+	// IsLeader returns whether this host is the leader
+	IsLeader() (bool, error)
 }
 
-// New returns a LeaderElector with the default service endpoint
-func New() *LeaderElector {
-	return &LeaderElector{
-		URL: defaultURL,
-	}
+type LeaderForLife struct {
+	isLeader atomic.Value
 }
 
-// NewWithURL returns a LeaderElector with a custom service endpoint URL
-func NewWithURL(URL string) *LeaderElector {
-	return &LeaderElector{
-		URL: URL,
-	}
+// New returns a Elector that uses the operator-sdk's leader for life elector
+func New(ctx context.Context, logger hclog.Logger, clientset kubernetes.Interface) *LeaderForLife {
+	le := &LeaderForLife{}
+	le.isLeader.Store(false)
+
+	go func() {
+		// This function blocks until this replica becomes the "leader", which
+		// means it creates a ConfigMap with an ownerref. Another replica can
+		// become the leader when the current leader replica stops running, and
+		// the Kubernetes garbage collector deletes the vault-k8s-leader
+		// ConfigMap.
+		err := operator_leader.Become(ctx, "vault-k8s-leader")
+		if err != nil {
+			logger.Error("trouble becoming leader:", "error", err)
+			return
```

Comment:
Given that we don't restart the `Become` attempt, could we end up in a situation where none of the replicas are trying to become the leader, if they all error here? Seems like maybe we should have this error feed back into some sort of retry logic to guard against that.

Reply:
@benashz Hmm, I added retries on error in https://github.com/hashicorp/vault-k8s/pull/271/commits/75258f728fb7e49cff1322925b15ceae4110af7a, I wonder if you're looking at an old diff?

Exiting right away is an interesting idea too.

Reply:
The more I think about this, the more I think it would be better for this to retry a few times, then if Become() is still failing, throw a signal or something that would do a system exit non-zero. Otherwise folks would need to be watching the logs to know there was a problem with the replica.

Reply:
@benashz @tomhjp added retries then exit non-zero in b73527f.

Reply:
I like that line of thinking 👍